engineapi/proxy.go

```diff
@@ -1,168 +1,421 @@
+
+				h.logger.Warnf("%v instance manager no longer exist for engine client proxy", pool.instanceManagerName)
+				for _, client := range pool.clients {
+					engineClientProxy := client.getEngineClientProxy()
+					log := engineClientProxy.getLogger()
+					log.Debug("Cleanup engine client proxy")
+
+					engineClientProxy.setNextGC()
+				}
+				continue
+			}
+
+			for clientUUID, client := range pool.clients {
+				engineClientProxy := client.getEngineClientProxy()
+				if !engineClientProxy.shouldResync() {
+					continue
+				}
+
+				log := engineClientProxy.getLogger()
+
+				if engineClientProxy.shouldDelete() {
+					log.Debug("Deleting engine client proxy")
+					engineClientProxy.setNextGC()
+
+					if err := h.NewEngineClientProxy(im); err != nil {
+						log.WithError(err).Debug("Failed to replace with new engine client proxy")
+					}
+
+					engineClientProxy.setNextResync(h.resyncPeriod)
+					break
+				}
+
+				h.gc(instanceManagerImage, clientUUID)
+
+				if engineClientProxy.shouldCreate(im) {
+					log.Debug("Creating engine client proxy")
+
+					if err := engineClientProxy.start(im, h.logger, h.ds, h.stopCh); err != nil {
+						log.Warn("Failed to start engine client proxy")
+					}
+				}
+
+				engineClientProxy.setNextResync(h.resyncPeriod)
+			}
+
+			exceedMaxConnection := true
+			for _, client := range pool.clients {
+				engineClientProxy := client.getEngineClientProxy()
+				if engineClientProxy.getInCall() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(pool.clients) != 0 && exceedMaxConnection {
+				log := getLoggerForProxyHandler(h.logger, im)
+				log.Debugf("Exceed engine client proxy max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				if err := h.NewEngineClientProxy(im); err != nil {
+					log.WithError(err).Debug("Failed to add more engine client proxy")
+				}
+			}
+		}
+	}
+}
+
+func (h *EngineClientProxyHandler) gc(instanceManagerImage string, clientUUID string) {
+	client, exist := h.getEngineClientProxy(instanceManagerImage, clientUUID)
+	if !exist {
+		h.logger.Debugf("Engine client proxy(%v) for %v is already gone", clientUUID, instanceManagerImage)
+		return
+	}
+
+	log := client.getLogger()
+
+	nextGC := client.getNextGC()
+	now := time.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	msg := ""
+	if client.getInCall() != 0 {
+		msg = fmt.Sprintf("Cannot stop engine client proxy with %v still in call", client.getInCall())
+		if client.getGCRetry() < ProxyGCMaxRetry {
+			log.Warnf(msg)
+			client.setNextGC()
+			return
+		}
+	}
+
+	log.Debug("Stopping engine client proxy")
+	if err := client.stop(); err != nil {
+		msg = "Failed to stop engine client proxy"
+		if client.getGCRetry() < ProxyGCMaxRetry {
+			log.WithError(err).Warnf(msg)
+			client.setNextGC()
+			return
+		}
+	}
+
+	if msg != "" {
+		log.Warn("Exceed max retry limit, force delete engine client proxy")
+	}
+
+	h.lock.Lock()
```

Comment:
note: Need to ensure all GC-marked clients will not be able to use by other requests. Otherwise, lock here is still not atomic, because a stopping client still has a chance to use.

Reply:
https://github.com/longhorn/longhorn-manager/pull/1338#discussion_r880394409

Reply:
`GC-marked clients will not be able to use by other requests` should be guaranteed by the lock inside the client rather than `h.lock`